src/MongoDB.Driver.Core/Core/Clusters/ClusterDescription.cs

```diff
@@ -38,6 +38,40 @@ internal static ClusterDescription CreateInitial(ClusterId clusterId, ClusterCon
                 ClusterType.Unknown,
                 Enumerable.Empty<ServerDescription>());
         }
+
+        // private static methods
+        private static TimeSpan? CalculateLogicalSessionTimeout(ClusterConnectionMode connectionMode, IEnumerable<ServerDescription> servers)
+        {
+            TimeSpan? logicalSessionTimeout = null;
+
+            foreach (var server in SelectServersThatDetermineWhetherSessionsAreSupported(connectionMode, servers))
```

Comment:
The call to SelectServersThatDetermineWhetherSessionsAreSupported is new.

There was a small bug in how the LogicalSessionTimeout was being computed for clusters with direct ConnectionMode.